plugins/inputs/cpufreq/cpufreq.go

```diff
@@ -0,0 +1,158 @@
+// +build linux
+
+package cpufreq
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+const defaultHostSys = "/sys"
+
+type CPUFreq struct {
+	Log             telegraf.Logger `toml:"-"`
+	PathSysfs       string          `toml:"host_sys"`
+	GatherThrottles bool            `toml:"gather_throttles"`
+	cpus            []cpu
+}
+
+type cpu struct {
+	id    string
+	path  string
+	props map[string]*os.File
+}
+
+var sampleConfig = `
+  ## path for sysfs filesystem.
+  ## See https://www.kernel.org/doc/Documentation/filesystems/sysfs.txt
+  ## Defaults:
+  # host_sys = "/sys"
+
+  ## Gather cpu throttles per core
+  ## Defaults:
+  # gather_throttles = false
+`
+
+func (g *CPUFreq) SampleConfig() string {
+	return sampleConfig
+}
+
+func (g *CPUFreq) Description() string {
+	return "Collects current cpu's frequencies"
+}
+
+func (g *CPUFreq) Init() error {
+	if g.PathSysfs == "" {
+		g.PathSysfs = defaultHostSys
+	}
+
+	cpus, err := g.discoverCpus()
+	if err != nil {
+		return err
+	}
+	g.cpus = cpus
+
+	return nil
+}
+
+func (g *CPUFreq) Gather(acc telegraf.Accumulator) error {
+	for _, cpu := range g.cpus {
+		fields := make(map[string]interface{})
+		tags := map[string]string{"cpu": cpu.id}
+
+		failed := false
+		for name, fd := range cpu.props {
+			v, err := readUintFromFile(fd)
+			if err != nil {
+				acc.AddError(err)
+				failed = true
+				break
+			}
+
+			fields[name] = v
+		}
+
+		if !failed {
+			acc.AddFields("cpufreq", fields, tags)
+		}
+	}
+
+	return nil
+}
+
+func (g *CPUFreq) discoverCpus() ([]cpu, error) {
+	var cpus []cpu
+
+	cpuDirs, err := filepath.Glob(path.Join(g.PathSysfs, "devices/system/cpu/cpu[0-9]*"))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(cpuDirs) == 0 {
+		return nil, fmt.Errorf("no CPUs detected at: %s", path.Join(g.PathSysfs, "devices/system/cpu/cpu[0-9]*"))
+	}
+
+	for _, dir := range cpuDirs {
+		_, cpuName := filepath.Split(dir)
+		cpuNum := strings.TrimPrefix(cpuName, "cpu")
+
+		if _, err := os.Stat(filepath.Join(dir, "cpufreq")); os.IsNotExist(err) {
+			return nil, fmt.Errorf("CPUFreq subsystem not available for cpu %s", cpuNum)
```

Comment:
Maybe you can write something about this subsystem (and how to enable it) in `README.md`?

Please see `README.md` for `intel_powerstat` plugin.

Reply:
CPUFreq must be enabled during build time of the kernel, though it is enabled in all mainstream distributions AFAIK. Is it it sufficient to say that it must be supported by the user's kernel (along with a description of the subsystem)?